hypershift-operator/controllers/scheduler/dedicated_content_serving_nodes.go

```diff
@@ -0,0 +1,204 @@
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	ControlPlaneTaint                              = "hypershift.openshift.io/control-plane"
+	ControlPlaneServingComponentTaint              = "hypershift.openshift.io/control-plane-serving-component"
+	HostedClusterTaint                             = "hypershift.openshift.io/cluster"
+	ControlPlaneServingComponentAvailableNodeTaint = "hypershift.openshift.io/control-plane-serving-component-available"
+
+	ControlPlaneServingComponentLabel = "hypershift.openshift.io/control-plane-serving-component"
+
+	HostedClusterNameLabel      = "hypershift.openshift.io/cluster-name"
+	HostedClusterNamespaceLabel = "hypershift.openshift.io/cluster-namespace"
+)
+
+type DedicatedServingComponentNodeReaper struct {
+	client.Client
+}
+
+func (r *DedicatedServingComponentNodeReaper) SetupWithManager(mgr ctrl.Manager) error {
+	servingComponentPredicate, err := predicate.LabelSelectorPredicate(metav1.LabelSelector{MatchLabels: map[string]string{hyperv1.RequestServingComponentLabel: "true"}})
+	if err != nil {
+		return err
+	}
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Node{}).WithEventFilter(servingComponentPredicate).Named("DedicatedServingComponentNodeReaper")
+	return builder.Complete(r)
+}
+
+func (r *DedicatedServingComponentNodeReaper) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx, "node", req.Name)
+	node := &corev1.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("node not found, aborting reconcile", "name", req.NamespacedName.String())
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get node %q: %w", req.NamespacedName.String(), err)
+	}
+	log.Info("Reconciling node")
+
+	if _, hasServingComponentLabel := node.Labels[hyperv1.RequestServingComponentLabel]; !hasServingComponentLabel {
+		log.Info("Skipping node because it doesn't have the control plane serving component label")
+		return ctrl.Result{}, nil
+	}
+
+	if _, hasHostedClusterLabel := node.Labels[hyperv1.HostedClusterLabel]; !hasHostedClusterLabel {
+		log.Info("Skipping node because it has not been allocated to a hosted cluster")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("Node has been allocated to a hosted cluster, checking whether hosted cluster still exists.")
+	name := node.Labels[HostedClusterNameLabel]
+	namespace := node.Labels[HostedClusterNamespaceLabel]
+	hc := &hyperv1.HostedCluster{}
+	if err := r.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, hc); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to get hosted cluster %s/%s: %w", namespace, name, err)
+		}
+		log.Info("The hosted cluster is not found. Deleting node.")
```

Comment:
TODO: Consider this vs set a condition in the Node instead.

Reply:
As discussed, we will consider conditions when looking to integrate this into the API